doc/feature/spikeReport.md

```diff
@@ -0,0 +1,450 @@
+  access.
+
+* Allow partial data loading for file access.
+
+* Allow the selection of a subset of the cells present in the data source to be
+  reported. In stream-based reports this information should be made available
+  to the sender to allow filtering at the source. In the ZeroEQ implementation,
+  the same channel needed for communicating the desired target could be used
+  for the handshaking required to ensure that the first spikes are not missed.
+
+* As few stream related details as possible should leak. This means that
+  it should be easy to write client code that works the same for both file
+  based and stream based reports without special conditions for stream based
+  reports.
+
+* For stream reports the user will be able to:
+ * Retrieve all the data until the end of stream in a simple loop without doing
+   any active wait (blocking calls are allowed).
+ * Know if the end of stream has been reached.
+ * Get all the data received so far without locking.
+ * Get information about the latest timestamp reached by the source in the
+   sender side even when there are no spikes at all (currently if the spike
+   report is completely empty it's not possible to make any progress until the
+   producer reaches the end of the report).
+
+* When data for a time interval is requested, the time intervals are always
+  closed on the left and open on the right.
+
+## API
+
+The classes presented below are not thread-safe except for const access.
+
+### Spikes container
+
+    Spikes
+    {
+    public:
+        Spikes();
+        ~Spikes();
+        /*
+         * Provided to allow Boost.Python to use return by value policies.
+         * Constant complexity.
+         */
+        Spikes( const Spikes& rhs );
+        /*
+         * Provided to allow Boost.Python to use return by value policies.
+         * Constant complexity.
+         */
+        Spikes& operator=( const Spikes& rhs );
+
+        class const_iterator; // implementation defined
+
+        const_iterator begin() const;
+        const_iterator end() const;
+        bool empty() const;
+        size_t size() const;
+        /* Get the spike at the nth position. No range check is performed */
+        const std::pair< float, uint32_t >& operator[]( size_t index ) const;
+
+        /*
+         * Get the start of the time window in milliseconds.
+         *
+         * UNDEFINED_TIMESTAMP if the container cannot be referred to any time
+         * window and it is empty. Otherwise, a time smaller or equal than the
+         * smallest spike time in the container.
+         *
+         * @see SpikesReport::getSpikes.
+         */
+        float getStartTime() const;
+
+        /*
+         * Get the end of the time window in milliseconds.
+         *
+         * UNDEFINED_TIMESTAMP if the container cannot be referred to any time
+         * window and it is empty. Otherwise, a time greater than the highest
+         * spike time in the container.
```

Comment:
Consistency with the specification of the interval passed to getSpikes as open on the right, I think it's mentioned in the issues (we may add it otherwise). If you call getSpikes using the values returned by this container you will get back the same content again. Otherwise it would be shrinking. I think it's important to always use the same time of intervals ([start, end)), because it makes reasoning about interval algebra easier. This specification also prevents having things like [t, t) which is wrong, and [t, t], which is not only inconsistent with getSpikes() but may also be problematic because it has 0 duration but still may contain events. In the specification of getSpikes it says (maybe not so clearly) that the end time must be either undefined or greater than the start time.